synapse/handlers/room.py

```diff
@@ -902,7 +907,13 @@ def filter_evts(events):
         state = yield self.state_store.get_state_for_events(
             [last_event_id], state_filter=state_filter
         )
-        results["state"] = list(state[last_event_id].values())
+
+        # Apply the filter on state events.
+        state_events = list(state[last_event_id].values())
+        if event_filter:
+            state_events = event_filter.filter(state_events)
+
+        results["state"] = list(state_events)
```

Comment:
I don't think this second `list` is necessary?
```suggestion
        results["state"] = state_events
```